src/QirRuntime/public/IQuantumApi.hpp

```diff
@@ -0,0 +1,146 @@
+#pragma once
+
+#include <complex>
+
+#include "CoreTypes.hpp"
+
+namespace Microsoft
+{
+namespace Quantum
+{
+    struct IQuantumApi
+    {
+        virtual ~IQuantumApi() {}
+
+        // TODO: Only some simulators 'know' the state -- is it OK to have this method on IQuantumApi?
+        // The callback will be invoked on each basis vector (in the standard computational basis) in little-endian
+        // order until it returns `false` or the state is fully dumped.
+        typedef bool (*TGetStateCallback)(size_t /*basis vector*/, double /* amplitude Re*/, double /* amplitude Im*/);
+        virtual void GetState(TGetStateCallback callback) = 0;
+
+        virtual std::string DumpQubit(Qubit qubit) = 0;
+
+        // Qubit management
+        virtual Qubit AllocateQubit() = 0;
+        virtual void ReleaseQubit(Qubit qubit) = 0;
+
+        // Shortcuts
+        virtual void CNOT(Qubit control, Qubit target)
+        {
+            CX(control, target);
+        }
+        virtual void CX(Qubit control, Qubit target) = 0;
+        virtual void CY(Qubit control, Qubit target) = 0;
+        virtual void CZ(Qubit control, Qubit target) = 0;
+
+        // Elementary operatons
+        virtual void X(Qubit target) = 0;
+        virtual void Y(Qubit target) = 0;
+        virtual void Z(Qubit target) = 0;
+        virtual void H(Qubit target) = 0;
+        virtual void S(Qubit target) = 0;
+        virtual void T(Qubit target) = 0;
+        virtual void SWAP(Qubit target1, Qubit target2) = 0;
+        virtual void Clifford(CliffordId cliffordId, PauliId pauli, Qubit target) = 0;
+        // unitary is an array of columns: 2^n x 2^n. Columns are nullable, so for state prep use just the first column
+        virtual void Unitary(long numTargets, double** unitary, Qubit targets[]) = 0;

```

Comment:
This only makes sense for a small number of specific simulators. It doesn't belong in a base class.